sunxspex/sunxspex_fitting/fitter.py

```diff
@@ -0,0 +1,4931 @@
+
+import types
+import math as maths
+
+# for colour list cycling
+from itertools import cycle
+
+# for Hessian matrix calculation for the minimise errors
+import numdifftools as nd
+from scipy.linalg import LinAlgError
+import pickle
+
+# let's not have warnings for every /0 when we have 0 counts
+import warnings
+warnings.filterwarnings("ignore", category=RuntimeWarning)
+warnings.filterwarnings("ignore", category=np.VisibleDeprecationWarning)
+
+from sunxspex.sunxspex_fitting.rainbow_text import rainbow_text_lines
+from sunxspex.sunxspex_fitting.photon_models_for_fitting import (defined_photon_models, f_vth, thick_fn, thick_warm)
+from sunxspex.sunxspex_fitting.likelihoods import LogLikelihoods
+from sunxspex.sunxspex_fitting.data_loader import LoadSpec
+from sunxspex.sunxspex_fitting.instruments import rebin_any_array
+from sunxspex.sunxspex_fitting.parameter_handler import Parameters, isnumber
+
+
+__all__ = ["add_var", "del_var","add_photon_model", "del_photon_model", "SunXspex", "load"]
+
+DYNAMIC_FUNCTION_SOURCE = {}
+
+# models should return a photon spectrum with units photons s^-1 cm^-2 keV^-1
+def add_photon_model(function, overwrite=False):
+    """ Add user photon model to fitting namespace.
+
+    Takes a user defined function intended to be used as a model or model component when giving a
+    string to the SunXspex.model property. Puts defined_photon_models[function.__name__]=param_inputs
+    in `defined_photon_models` for it to be known to the fititng code. The energies argument must be
+    first and accept photon bins.
+
+    The given function needs to have parameters as arguments then \'energies\' as a keyword argument
+    where energies accepts the energy bin array. E.g.,
+
+    .. math::
+     gauss = = a$\cdot$e$^{-\frac{(energies - b)^{2}}{2 c^{2}}}$
+
+    would be
+     `gauss = lambda a, b, c, energies=None: a * np.exp(-((np.mean(energies, axis=1)-b)**2/(2*c**2)))`
+
+    Parameters
+    ----------
+    function : function
+            The function object.
+
+    overwrite : bool
+            Set True to overwrite a model that already exists. User needs to be explicit if they wish
+            to overwrite their models.
+            Default: False
+
+    Returns
+    -------
+    None.
+
+    Example
+    -------
+    from fitter import SunXspex, defined_photon_models, add_photon_model
+
+    # Define gaussian model (doesn't have to be a lambda function)
+    gauss = lambda a, b, c, energies=None: a * np.exp(-((np.mean(energies, axis=1)-b)**2/(2*c**2)))
+
+    # Add the gaussian model to fitter.py namespace
+    add_photon_model(gauss)
+
+    # Now can use it in fitting with string defined model. Will be plotted separately to the total model
+    Sx = SunXspex(pha_file=[...])
+    Sx.model = "gauss+gauss"
+    Sx.fit()
+    Sx.plot()
+    """
+    # if user wants to define any component model to use in the fitting and reference as a string
+    _glbls_cp, _dfs_cp = copy(globals()), copy(DYNAMIC_FUNCTION_SOURCE)
```

Comment:
I absolutely agree and I only wrote it like this to get things working properly. The issue comes when pickling the class, either for parallelisation or just to save out the session, where the function/variable/etc. cannot be used in another session. This does need changed and I have been working on a different solution where the dynamic user functions/variables/etc. are written to their own .py file then can always be loaded in from that location. This would allow the user to have far more power over their analysis with other benefits too. However, I haven't managed to implement this yet. Other solutions to get around this issue are also absolutely welcome.